Add Unit Tests for BPMN Steps

DiCanio opened the pull request:
Resolves #1

DiCanio committed:
Add Unit Tests for BPMN Steps

DiCanio commented:
Tests for the `AggregateResults` step have been omitted since this has to be touched anyway when working on #6.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,194 @@
+
+    @Test
+    public void testDoExecute_NoMeasureReference() {
+        when(execution.getVariable(Mockito.eq("task"))).thenReturn(task);
```
Please use ConstantsBase.BPMN_EXECUTION_VARIABLE_TASK instead of "task" and why is Mockito.eq necessary?

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,194 @@
+    @Test
+    public void testDoExecute_NoMeasureReference() {
+        when(execution.getVariable(Mockito.eq("task"))).thenReturn(task);
+        when(taskHelper.getFirstInputParameterReferenceValue(task, ConstantsFeasibility.CODESYSTEM_FEASIBILITY,
```
Please add static import to enhance readability.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,194 @@
+                .thenReturn(Optional.of(measureRef));
+        when(clientProvider.getLocalWebserviceClient())
+                .thenReturn(localWebserviceClient);
+        when(clientProvider.getRemoteWebserviceClient(Mockito.anyString()))
```
It is either the local or the remote webservice client but not both. Please test both scenarios.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,194 @@
+                .thenReturn(localWebserviceClient);
+        when(clientProvider.getRemoteWebserviceClient(Mockito.anyString()))
+                .thenReturn(localWebserviceClient);
+        when(localWebserviceClient.searchWithStrictHandling(Mockito.any(), Mockito.anyMap()))
```
Please use real values here. It's not hard to build them.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,194 @@
+    public void testDoExecute_BundleWithTooFewResultEntries() {
+        final Reference measureRef = new Reference("http://localhost/Measure/id-151003");
+
+        final Bundle measureOnlyBundle = new Bundle(new Enumeration<>(new Bundle.BundleTypeEnumFactory()));
```
Is it necessary to create the Bundle with the BundleTypeEnumFactory?
The name measureOnlyBundle is missleading, because the Bundle is empty.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,194 @@
+        final Reference measureRef = new Reference("http://localhost/Measure/id-151003");
+
+        final Bundle measureOnlyBundle = new Bundle(new Enumeration<>(new Bundle.BundleTypeEnumFactory()));
+        measureOnlyBundle.setType(Bundle.BundleType.BATCHRESPONSE);
```
The Bundle type is never tested. So why do you set it. It's even wrong, because it should be searchset.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,194 @@
+
+        final Bundle measureOnlyBundle = new Bundle(new Enumeration<>(new Bundle.BundleTypeEnumFactory()));
+        measureOnlyBundle.setType(Bundle.BundleType.BATCHRESPONSE);
+        measureOnlyBundle.addEntry(patientEntry);
```
You can use fluent style here: bundle.addEntry().setResource(...)

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,194 @@
+        patientEntry.setResource(new Patient().setId("id-170524"));
+
+        final Bundle.BundleEntryComponent measureEntry = new Bundle.BundleEntryComponent();
+        measureEntry.setId("id-151003");
```
BundleEntry.id is never used.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,194 @@
+
+        final Bundle.BundleEntryComponent measureEntry = new Bundle.BundleEntryComponent();
+        measureEntry.setId("id-151003");
+        measureEntry.setResource(new Measure().setId("id-170418"));
```
A non-Measure first entry is sufficient.

DiCanio replied:
No it's not. If we were about to test it like this then the `bundle.getEntry().size() < 2` check within the tested class would return true and we would end up in another execution branch.

alexanderkiel replied:
Ok you are right. But please use two Patients to keep things simpler.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,194 @@
+                .thenReturn(localWebserviceClient);
+        when(localWebserviceClient.searchWithStrictHandling(Mockito.any(), Mockito.anyMap()))
+                .thenReturn(measureOnlyBundle);
+        when(localWebserviceClient.getBaseUrl()).thenReturn("http://localhost");
```
Is only used in warn logging. But not in this positive case.

alexanderkiel reviewed (changes requested):
Please use the comment on the first test to improve all other tests. I will do a second review after that.

DiCanio committed:
Use Predefined Task Variable

DiCanio committed:
Use Static Imports Where Appropriate

DiCanio committed:
Improve Tests for DownloadFeasibilityResources

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,303 @@
+                .thenReturn(Optional.of(measureRef));
+        when(clientProvider.getLocalWebserviceClient())
+                .thenReturn(webserviceClient);
+        when(webserviceClient.searchWithStrictHandling(ArgumentMatchers.<Class<MeasureReport>>any(), eq(createSearchQueryParts(measureId))))
```
Can simply be `Measure.class, createSearchQueryParts(measureId)`.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,303 @@
+    @Test
+    public void testDoExecute_BundleWithTooFewResultEntriesFromLocal() {
+        final String measureId = "id-151003";
+        final Reference measureRef = new Reference("/Measure/" + measureId);
```
A local Reference doesn't start with a slash. It is just `<type>/<id>`

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,303 @@
+                .thenReturn(webserviceClient);
+        when(webserviceClient.searchWithStrictHandling(ArgumentMatchers.<Class<MeasureReport>>any(), eq(createSearchQueryParts(measureId))))
+                .thenReturn(new Bundle());
+        when(webserviceClient.getBaseUrl()).thenReturn("http://localhost");
```
Can be removed.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,303 @@
+                .thenReturn(webserviceClient);
+        when(webserviceClient.searchWithStrictHandling(ArgumentMatchers.<Class<MeasureReport>>any(), eq(createSearchQueryParts(measureId))))
+                .thenReturn(new Bundle());
+        when(webserviceClient.getBaseUrl()).thenReturn("http://remote.host");
```
Can be removed.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,303 @@
+                .thenReturn(Optional.of(measureRef));
+        when(clientProvider.getRemoteWebserviceClient("http://remote.host"))
+                .thenReturn(webserviceClient);
+        when(webserviceClient.searchWithStrictHandling(ArgumentMatchers.<Class<MeasureReport>>any(), eq(createSearchQueryParts(measureId))))
```
Can simply be `Measure.class, createSearchQueryParts(measureId)`.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,303 @@
+        patientEntry.setResource(new Patient().setId("id-170524"));
+
+        final Bundle.BundleEntryComponent measureEntry = new Bundle.BundleEntryComponent();
+        measureEntry.setResource(new Measure().setId("id-170418"));
```
Please just use two Patients, to be less specific here.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,303 @@
+    }
+
+    @Test
+    public void testDoExecute_FirstBundleEntryIsNoMeasureFromLocal() {
```
It is sufficient to test the local branch once. You did this already above. All other branches should be remote, because the feasibility reasources are always downloaded from remote.

alexanderkiel reviewed (changes requested):
Please test the local branch only once.

DiCanio committed:
Introduce Enhanced Client Provider

The enhanced version encapsulates the creation of
a specific FHIR webservice client (local / remote)
based on a given IdType.
This helps with different testing scenarios.

DiCanio committed:
Improve Enhanced Client Provider

alexanderkiel commented on codex-process-feasibility/src/main/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResources.java:
```diff
@@ -44,8 +44,8 @@ protected void doExecute(DelegateExecution execution) {
-        FhirWebserviceClient client = getWebserviceClient(measureId);
-        Bundle bundle = getMeasureAndLibrary(measureId, client);
+        Bundle bundle = getMeasureAndLibrary(measureId,
+                ((EnhancedFhirWebserviceClientProvider) getFhirWebserviceClientProvider()).getWebserviceClient(measureId));
```
Would put this pack into the variable `FhirWebserviceClient client = ` because it's a bit to much code here.

alexanderkiel commented on codex-process-feasibility/src/main/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResources.java:
```diff
@@ -28,7 +28,7 @@
     private final OrganizationProvider organizationProvider;
 
-    public DownloadFeasibilityResources(FhirWebserviceClientProvider clientProvider, TaskHelper taskHelper, OrganizationProvider organizationProvider) {
+    public DownloadFeasibilityResources(EnhancedFhirWebserviceClientProvider clientProvider, TaskHelper taskHelper, OrganizationProvider organizationProvider) {
```
Please break before 120 chars.

alexanderkiel commented on codex-process-feasibility/src/main/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/spring/config/FeasibilityConfig.java:
```diff
@@ -16,29 +18,43 @@
 
-    @Autowired
-    private FhirWebserviceClientProvider fhirClientProvider;
+    private final FhirWebserviceClientProvider fhirClientProvider;
```
Please remove empty lines between the fields.

alexanderkiel commented on codex-process-feasibility/src/main/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/spring/config/FeasibilityConfig.java:
```diff
@@ -16,29 +18,43 @@
 import org.springframework.context.annotation.Configuration;
 
 @Configuration
 public class FeasibilityConfig {
```
Run the whole process in Docker test setup to ensure that the new Spring config actually works.

alexanderkiel commented on codex-process-feasibility/src/main/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/EnhancedFhirWebserviceClientProvider.java:
```diff
@@ -0,0 +1,18 @@
+public interface EnhancedFhirWebserviceClientProvider extends FhirWebserviceClientProvider {
+
+    /**
+     * Gets a {@link FhirWebserviceClient} for local or remote use based on the given reference.
```
Please add a newline after the first sentence.

alexanderkiel commented on codex-process-feasibility/src/main/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/EnhancedFhirWebserviceClientProviderImpl.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @Override
+    public FhirWebserviceClient getRemoteWebserviceClient(String s, String s1) {
```
Please use the real parameter names here.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,223 @@
+    @Test
+    public void testDoExecute_BundleWithTooFewResultEntries() {
+        final String measureId = "id-151003";
+        final IdType measureRefId = new IdType("http://remote.host/Measure/" + measureId);
```
A local ID will do now. So please remove `http://remote.host/` to make things simpler.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,223 @@
+
+    @Test
+    public void testDoExecute_BundleWithTooFewResultEntries() {
+        final String measureId = "id-151003";
```
Please don't use local finals. IMHO it's not worth the fuzz.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,223 @@
+        final Bundle.BundleEntryComponent patientEntryB = new Bundle.BundleEntryComponent()
+                .setResource(new Patient().setId("id-123456"));
+
+        final Bundle measureOnlyBundle = new Bundle()
```
```java
Bundle bundle = new Bundle();
bundle.addEntry().setResource(new Patient().setId("foo"));
bundle.addEntry().setResource(new Patient().setId("foo"));
```

is sufficient. I would not use a special ID here, because all it needs is a non-empty resource. Please replace all other ID's this way.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,223 @@
+    }
+
+    @Test
+    public void testDoExecuteLocal() throws Exception {
```
We only need one case, not a local and a remote one.

alexanderkiel reviewed (changes requested):
More remarks on DownloadFeasibilityResourcesTest. Will look at the other tests after DownloadFeasibilityResourcesTest is sorted out.

DiCanio committed:
Improve Tests

DiCanio committed:
Line Breaks and Readability

DiCanio committed:
JavaDoc Improvements

DiCanio committed:
Use Real Parameter Names

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,187 @@
+
+    @Test
+    public void testDoExecute_BundleWithTooFewResultEntries() {
+        String measureId = "foo";
```
This ID shouldn't be foo because it is relevant. It would be even better to put it in a constant. So you save a few lines of code.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/service/DownloadFeasibilityResourcesTest.java:
```diff
@@ -0,0 +1,187 @@
+        Bundle.BundleEntryComponent patientEntryB = new Bundle.BundleEntryComponent()
+                .setResource(new Patient().setId("foo"));
+
+        Bundle measureOnlyBundle = new Bundle()
```
This is not a measureOnlyBundle :-) And please use the builder-style pattern to add the entries. This saves many lines of code. I know the HAPI classes are not desinged very well because they are their own builders. But I would use that nevertheless, because it's shorter and I've seen it many times.

DiCanio replied:
Made some changes. Hope this fits your requirements.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/EnhancedFhirWebserviceClientProviderImplTest.java:
```diff
@@ -0,0 +1,76 @@
+
+    @Test
+    public void testGetWebserviceClient_Local() {
+        final IdType idType = new IdType("Something/id-123456");
```
Please remove the finals

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/EnhancedFhirWebserviceClientProviderImplTest.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Test
+    public void testGetRemoteWebserviceClient() {
```
Should be three tests. Should not use verify. Please test it like the first tests.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/EnhancedFhirWebserviceClientProviderImplTest.java:
```diff
@@ -0,0 +1,76 @@
+    @Test
+    public void testGetLocalBaseUrl() {
+        enhancedFhirWebserviceClientProvider.getLocalBaseUrl();
+        verify(clientProvider).getLocalBaseUrl();
```
Please don't use verify.

alexanderkiel commented on codex-process-feasibility/src/test/java/de/netzwerk_universitaetsmedizin/codex/processes/feasibility/EnhancedFhirWebserviceClientProviderImplTest.java:
```diff
@@ -0,0 +1,76 @@
+    private EnhancedFhirWebserviceClientProviderImpl enhancedFhirWebserviceClientProvider;
+
+    @Test
+    public void testGetWebserviceClient_Local() {
```
There is a second local case when the baseURl is the same as the providers baseURL. Please add another test.

alexanderkiel reviewed (changes requested):
There are still things to do at DownloadFeasibilityResourcesTest and EnhancedFhirWebserviceClientProviderImplTest. Have not looked at the other tests yet.

DiCanio committed:
[WIP] Fix Spring Configuration

DiCanio committed:
[WIP 2] Fix Spring Configuration

DiCanio committed:
Further Improvements

alexanderkiel committed:
Fix DI

alexanderkiel committed:
Improve DownloadFeasibilityResourcesTest

DiCanio committed:
Align Remaining Tests Regarding Improvements

alexanderkiel committed:
Some Formatting Stuff
